cmd/influxd/inspect/report_tsi/report_tsi.go

```
@@ -0,0 +1,481 @@
+// Cardinality struct and methods
+type cardinality struct {
+	name  []byte
+	short []uint32
+	set   *tsdb.SeriesIDSet
+}
+
+func (c *cardinality) add(x uint64) {
+	if c.set != nil {
+		c.set.AddNoLock(x)
+		return
+	}
+
+	c.short = append(c.short, uint32(x)) // Series IDs never get beyond 2^32
+
+	// Cheaper to store in bitmap.
+	if len(c.short) > useBitmapN {
+		c.set = tsdb.NewSeriesIDSet()
+		for i := 0; i < len(c.short); i++ {
+			c.set.AddNoLock(uint64(c.short[i]))
+		}
+		c.short = nil
+		return
+	}
+}
+
+func (c *cardinality) cardinality() int64 {
+	if c == nil || (c.short == nil && c.set == nil) {
+		return 0
+	}
+
+	if c.short != nil {
+		return int64(len(c.short))
+	}
+	return int64(c.set.Cardinality())
+}
+
+type cardinalities []*cardinality
+
+func (a cardinalities) Len() int           { return len(a) }
+func (a cardinalities) Less(i, j int) bool { return a[i].cardinality() < a[j].cardinality() }
+func (a cardinalities) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+
+func (report *reportTSI) cardinalityByMeasurement(shardID uint64) error {
+	idx := report.shardIdxs[shardID]
+	itr, err := idx.MeasurementIterator()
+	if err != nil {
+		return err
+	} else if itr == nil {
+		return nil
+	}
+	defer itr.Close()
+
+OUTER:
+	for {
+		name, err := itr.Next()
+		if err != nil {
+			return err
+		} else if name == nil {
+			break OUTER
+		}
+
+		// Get series ID set to track cardinality under measurement.
+		c, ok := report.cardinalities[shardID][string(name)]
+		if !ok {
+			c = &cardinality{name: name}
+			report.cardinalities[shardID][string(name)] = c
+		}
+
+		sitr, err := idx.MeasurementSeriesIDIterator(name)
+		if err != nil {
+			return err
+		} else if sitr == nil {
+			continue
+		}
+
+		var e tsdb.SeriesIDElem
+		for e, err = sitr.Next(); err == nil && e.SeriesID != 0; e, err = sitr.Next() {
+			if e.SeriesID > math.MaxUint32 {
+				panic(fmt.Sprintf("series ID is too large: %d (max %d). Corrupted series file?", e.SeriesID, uint32(math.MaxUint32)))
```

Why `panic` here instead of returning an `err`?

---

Hmm, I just went with what was already in the 1.x version. I think an error would be more appropriate here.